assignments/HW4.md

```diff
@@ -0,0 +1,194 @@
+---
+layout: page
+title: HW04 Microservices
+permalink: /assignments/hw4
+parent: Assignments
+---
+
+**17-313: Foundations of Software Engineering**
+
+# Homework 4: Microservices
+
+The key learning goals of this homework are that you are able to:
+
+- Train and deploy a machine learning model using microservices.
+
+- Decide whether and when to use microservices, understanding benefits and limitations.
+
+- Write an effective design document.
+
+- Describe the architecture of a system using a suitable notation, using architectural views to represent software along different perspectives.
+
+- Assess and choose between alternative architectural decisions.
+
+- Engage meaningfully with questions of engineering ethics.
+
+## Project context
+
+You and your team are hard at work on the new graduate admissions system for CMU. During your most recent sprint planning meeting, your CMU technical contact bursts into the room, shouting, “Machine learning! Microservices!” When pressed, your newly buzzword-enamored boss explains: the population of applicants is too large, and increasing annually. Admissions is taking too much time and effort. Fortunately, CMU has been collecting student performance data. They continue: “Can’t your team make use of all that fancy new machine learning to help predict which applicants are more likely to succeed, helping the admissions process along? And since we’re checking off buzzwords, can’t you implement it as a microservice?”
+
+You are correctly skeptical about the effort required to pivot the product in this new direction, but fortunately the somewhat disconnected microservice demand gives you a way to explore the idea without requiring a major up front change in the existing codebase. Instead, you’ll (1) develop this potential feature in isolation by training a machine learning model that predicts potential applicant success and deploying it as a microservice, and then (2) reason about, document, and choose between alternative designs that would allow you to integrate it into the existing codebase. Finally, you will (3) individually consider the ethical implications of the feature, and what questions you truly need to ask in QA to satisfy possible ethical concerns.
```

Comment:
`you’ll (1) develop this potential feature in isolation by training a machine learning model`

I think that we're providing them with a pretrained model? @rohanpadhye @namdy0429 @MichaelHilton

Reply:
`(2) reason about, document, and choose between alternative designs that would allow you to integrate it into the existing codebase`

I think that the design space is much more restricted compared to last year, especially given the starter code and need to communicate between Java and Python. Perhaps we can highlight that we're really asking them to design/document/reason about the API between the ML component and the big ole' monolith?

Reply:
1) The starter code, like last year, provides a default model. I thought the task was to create a better-performing model (perhaps by using a decision tree classifier) and create APIs for that.

2) The API between the ML component and the existing webapp will be documented by Swagger as part of the 1st deliverable. The alternative design is to integrate it as part of the monolith, and the students would discuss tradeoffs of that.

Reply:
Instructions for a microservice design doc explicated in the design doc section.